File: packages/adapters/hermes/test/children/broken_sends.py
"""The counter-example, committed on purpose.

This is ``wemessage_wire.py --standalone`` with the one change a stranger
makes first: it decides drafting is a formality and puts the text on the wire
itself, in a frame it invented, because the protocol has no such frame to
borrow.

It does not work, and WHERE it stops is the interesting part. The gateway does
not deliver the message and then ask forgiveness. The frame never parses, so
it is refused at the socket and audited as ``adapter.no-send-frame`` — never
``gate.denied``, because it did not reach a gate, an approval queue or a
human. ``FRAME_SPECS`` has nine entries and none of them puts text on a phone.

Note what this file had to DO to get here. ``WireClient._emit`` refuses any
type outside :data:`AGENT_FRAME_TYPES`, so a ``send`` frame cannot be emitted
through the client at all; the only way to write one is to reach past the
chokepoint and touch the socket directly, which is what the marked block
below does. That is the whole argument for having a chokepoint: the mistake is
still possible, but it is no longer possible by accident.

Run it and read the badge::

    npx @wemessage/adapter-testkit --cmd "python broken_sends.py"
    NOT CONFORMANT v1 - broken-sends-py

The Node twin of this file is ``packages/adapter-testkit/examples/broken-sends.mjs``.
Same mistake, same refusal, different language, which is the point: the
taxonomy is the daemon's, not the runtime's.
"""

from __future__ import annotations

import asyncio
import json
import os
import sys
import uuid
from typing import Any

import websockets

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "plugin"))

from wemessage_wire import WIRE_VERSION, idempotency_key, utc_now  # noqa: E402


async def session(url: str, adapter_id: str, token: str) -> None:
    async with websockets.connect(url) as websocket:

        async def write(frame_type: str, payload: dict[str, Any]) -> None:
            # No vocabulary check. That absence IS the bug being demonstrated.
            await websocket.send(
                json.dumps(
                    {
                        "v": WIRE_VERSION,
                        "id": str(uuid.uuid4()),
                        "type": frame_type,
                        "ts": utc_now(),
                        "payload": payload,
                    }
                )
            )

        await write("hello", {"adapterId": adapter_id, "token": token, "wire": 1})
        async for raw in websocket:
            try:
                frame = json.loads(raw)
            except (ValueError, TypeError):
                continue
            if not isinstance(frame, dict) or frame.get("v") != WIRE_VERSION:
                continue
            if frame.get("type") == "ping":
                await write("pong", {})
                continue
            if frame.get("type") != "draft.request":
                continue
            payload = frame.get("payload")
            if not isinstance(payload, dict):
                continue
            correlation = payload.get("correlation") or {}
            message = payload.get("message") or {}
            content = message.get("content") or {}

            # ── the change. Everything above is the reference adapter. ──
            await write(
                "send",
                {
                    "chatGuid": correlation.get("chatGuid"),
                    "body": content.get("text") or "",
                },
            )
            # ── and the frame above is why this file is NOT CONFORMANT. ──

            await write(
                "draft.submit",
                {
                    "correlation": correlation,
                    "idempotencyKey": idempotency_key(correlation),
                    "declined": True,
                },
            )


async def main() -> int:
    url = os.environ["WEMESSAGE_GATEWAY_URL"]
    adapter_id = os.environ["WEMESSAGE_ADAPTER_ID"]
    token = os.environ["WEMESSAGE_ADAPTER_TOKEN"]
    for _ in range(int(os.environ.get("WEMESSAGE_MAX_ATTEMPTS", "3"))):
        try:
            await session(url, adapter_id, token)
        except (OSError, websockets.WebSocketException):
            pass
    return 1


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: packages/adapter-testkit/examples/broken-sends.mjs
/**
 * The counter-example: `reference-adapter.mjs` with the one change a stranger
 * makes first. It is here so the refusal is demonstrable rather than merely
 * asserted, and so the error an adapter author will actually hit has a name
 * and a file attached to it.
 *
 * The change is the block marked below. This adapter decides that drafting is
 * a formality and puts the text on the wire itself, in a frame it invented,
 * because the protocol has no such frame to borrow.
 *
 * It does not work, and the interesting part is WHERE it stops. The gateway
 * does not deliver the message and then ask forgiveness; the frame never
 * parses, so it is refused at the socket and audited as
 * `adapter.no-send-frame`. It is never `gate.denied` — the gate is about
 * approvals, and this frame did not reach a gate, an approval queue, or a
 * human. The conformance kit fails checks 3 and 6 on it for the same reason
 * the daemon would drop it: `FRAME_SPECS` has nine entries and none of them
 * puts text on a phone. Approval is the human's, and it is not on this wire.
 *
 *   npx @wemessage/adapter-testkit --cmd "node broken-sends.mjs"
 *   NOT CONFORMANT v1 - broken-sends (2/6 failed)
 */
const env = process.env;
const adapterId = env.WEMESSAGE_ADAPTER_ID;
const token = env.WEMESSAGE_ADAPTER_TOKEN;
const maxAttempts = Number(env.WEMESSAGE_MAX_ATTEMPTS ?? 3);
const iso = () => new Date().toISOString();

const onFrame = (frame, emit) => {
  if (frame?.v !== 1) return;
  if (frame.type === 'ping') return emit('pong', {});
  if (frame.type !== 'draft.request') return;
  const { correlation, message } = frame.payload;
  const key = `bad:${correlation.inboundGuid ?? correlation.requestId}`;
  const text = message.content.text ?? '';

  /* ── the change. Everything above this line is the reference adapter. ── */
  emit('send', { chatGuid: correlation.chatGuid, body: text });
  /* ── and the frame above is why this file is NOT CONFORMANT. ────────── */

  emit('draft.submit', { correlation, idempotencyKey: key, declined: true });
};

const session = () =>
  new Promise((done) => {
    const socket = new WebSocket(env.WEMESSAGE_GATEWAY_URL);
    const emit = (type, payload) => {
      const id = crypto.randomUUID();
      socket.send(JSON.stringify({ v: 1, id, type, ts: iso(), payload }));
    };
    socket.onopen = () => emit('hello', { adapterId, token, wire: 1 });
    socket.onclose = socket.onerror = () => done();
    socket.onmessage = (event) => {
      try {
        onFrame(JSON.parse(String(event.data)), emit);
      } catch {
        /* malformed input is refused, never answered, and never fatal */
      }
    };
  });

for (let attempt = 1; attempt <= maxAttempts; attempt += 1) await session();
process.exit(1);
